### DISCUSSION TITLE
zh/Q-and-A/answer/question1

### DISCUSSION THREAD

**Discussion by giscus** [vuepress-theme-hope/giscus-discussions#913] [Announcements]
# zh/Q-and-A/answer/question1

为什么 unsigned int 的输出是-1呢 问题描述： 请问一下unsigned int等于-1时为什么不像unsigned char等于-1输出是个正数(255) 我尝试了用while设置了两个unsigned int变量一个等于0，一个等于1，然后让两个数每次循环加一，知道第二个数加到零为止，最后输出第一个数也是-1 我搜索了哔哩哔哩，CSD...

https://cygnomatic-controller.github.io/zh/Q-and-A/answer/question1.html

**Comment by GzqHerry** [comment]
因为printf中使用%d会默认c输出为一个int。如果想要输出unsigned int类型，则需要将%d改为%u.

**Comment by 477996850** [comment]
C语言中，unsigned int 是无符号整型，不支持负数。你将 -1 赋值给 unsigned int 时，实际上是进行了一种类型转换。
-1在二进制中通常表示为所有位都为1（例如，32位无符号整型中的 `0xFFFFFFFF`）。赋值时，C会将这个二进制值解释为一个无符号数，这样就变成了一个很大的正数。
你的代码输出 -1 是因为你使用了 %d 来打印 unsigned int，这个格式符是针对有符号整型的。为了正确显示无符号整型，你应该使用 %u

#include <stdio.h>
int main()
{
    unsigned int c = -1;
    printf("%u", c);  
    return 0;
}
这样会输出 4294967295（在32位系统上）。而对于 unsigned char，将 -1 赋值时，结果会是得255，因为它会对无符号字符的范围取模。

**Comment by MAGMADIMSUM-RTX** [comment]
int类型一般占4个字节，int变量存储在这四个字节中。在你的代码中，%d意味着告诉编译器：“我要print的这个数是一个int类型的，不管输入的数据是什么类型，你把它当int型的就好了”。编译器不会看c的数据类型，只看%d的类型以及传入的值（-1对应二进制数为0xFFFFFFFF，即使这个数存储在c中）。如果你使用%x，就会对应输出16进制数。（建议参考一下计算机中数据的存储，相信会对你有帮助）
